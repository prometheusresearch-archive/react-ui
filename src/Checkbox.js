/**
 * @copyright 2016-present, Prometheus Research, LLC
 */

import React from 'react';

export default class Checkbox extends React.Component {

  static stylesheet = {
    Root: 'div',
    Input: 'input',
    Label: 'label',
  };

  render() {
    let {value, label, ...props} = this.props;
    let {Root, Input, Label} = this.constructor.stylesheet;
    return (
      <Root>
        <Input
          {...props}
          type="checkbox"
          checked={value}
          onChange={this.onChange}
          />
        {label && <Label>{label}</Label>}
      </Root>
    );
  }

  onChange = e => {
    this.props.onChange(e.target.checked);
  };
}

