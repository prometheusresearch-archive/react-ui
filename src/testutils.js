/**
 * @copyright 2016-present, Prometheus Research, LLC
 */

import React from 'react';
import {isEqual} from 'lodash/lang';
import indent from 'indent-string';

import {
  findWithType as _findWithType,
  findAllWithType as _findAllWithType,
  getMountedInstance,
  findAll
} from 'react-shallow-testutils';

import {
  createRenderer as _createRenderer
} from 'react-addons-test-utils';

import _inspectElement, {
  inspectReactProp as _inspectReactProp
} from 'inspect-react-element';


import assert from 'power-assert';
import {stub, spy, mock} from 'sinon';

export {stub, spy, mock, assert};

class FormDataMock {

  append = spy();
}

let originalFormData = global.FormData;

export function mockFormData() {
  global.FormData = FormDataMock;
}

export function unmockFormData() {
  global.FormData = originalFormData;
}

function inspectReactProps(node) {
  return '__originalConfig' in node.props ?
    {...node.props.__originalConfig, children: node.props.children} :
    node.props || {};
}

function inspectReactProp(propName, propValue) {
  if (propName === '__originalConfig' || propName === 'key' || propName === 'ref') {
    return null;
  }
  return _inspectReactProp(propName, propValue);
}

export function inspectElement(node) {
  return _inspectElement(node, {inspectReactProps, inspectReactProp});
}

class ReactShallowRenderer {

  constructor() {
    this._renderer = _createRenderer();
  }

  render(element, context) {
    return this._renderer.render(element, context);
  }

  get instance() {
    return getMountedInstance(this._renderer);
  }

  get element() {
    return this._renderer.getRenderOutput();
  }

  findWithType(type) {
    return findWithType(this.element, type);
  }

  findAllWithType(type) {
    return findAllWithType(this.element, type);
  }

  assertElementWithType(type) {
    this.findWithType(type);
  }

  assertNoElementWithType(type) {
    assert(this.findAllWithType(type).length === 0);
  }


  findWithTypeProps(type, props) {
    return findWithTypeProps(this.element, type, props);
  }

  findAllWithTypeProps(type, props) {
    return findAllWithTypeProps(this.element, type, props);
  }

  assertElementWithTypeProps(type, props) {
    this.findWithTypeProps(type, props);
  }

  assertNoElementWithTypeProps(type, props) {
    assert(this.findAllWithTypeProps(type, props).length === 0);
  }


  findWithElement(element) {
    return findWithElement(this.element, element);
  }

  findAllWithElement(element) {
    return findAllWithElement(this.element, element);
  }

  assertElement(element) {
    this.findWithElement(element);
  }

  assertNoElement(element) {
    assert(this.findAllWithElement(element).length === 0);
  }

}

export function createRenderer() {
  return new ReactShallowRenderer();
}

export function findWithType(root, type) {
  let elements = findAll(root, _makeMatchWithType(type));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one partial match for element:
${indent(inspectElement(React.createElement(type)), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithType(root, type) {
  return findAll(root, _makeMatchWithType(type));
}

function _makeMatchWithType(type) {
  return function _matchWithType(element) {
    return element && element.type && (
      element.type === type ||
      element.type.Component === type
    );
  };
}

export function findWithTypeProps(root, type, props) {
  let elements = findAll(root, _makeMatchWithElement(type, props));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one (${elements.length}) partial match for element:
${indent(inspectElement(React.createElement(type, props)), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithTypeProps(root, type, props) {
  return findAll(root, _makeMatchWithElement(type, props));
}

export function findWithElement(root, spec) {
  spec = _normalizeReactElementSpec(spec);
  let elements = findAll(root, _makeMatchWithElement(spec.type, spec.props));
  if (elements.length !== 1) {
    throw new Error(`Did not find exactly one (${elements.length}) partial match for element:
${indent(inspectElement(spec), ' ', 7)}
${indent('In the rendered element tree:', ' ', 5)}
${indent(inspectElement(root), ' ', 7)}
    `);
  }
  return elements[0];
}

export function findAllWithElement(root, spec) {
  spec = _normalizeReactElementSpec(spec);
  return findAll(root, _makeMatchWithElement(spec.type, spec.props));
}

function _makeMatchWithElement(type, props) {
  let matchWithType = _makeMatchWithType(type);

  let matchWithProps = function(element) {
    for (let key in props) {
      if (!props.hasOwnProperty(key)) {
        continue;
      }
      if (!element.props.hasOwnProperty(key)) {
        continue;
      }
      if (key === '__originalConfig') {
        continue;
      }
      if (!isEqual(element.props[key], props[key])) {
        return false;
      }
    }
    return true;
  };
  return function _matchWithElement(element) {
    return element && matchWithType(element) && matchWithProps(element);
  };
}

function _normalizeReactElementSpec(spec) {
  let {props: {__originalConfig = null}} = spec;
  if (__originalConfig) {
    return {...spec, props: __originalConfig};
  } else {
    return spec;
  }
}

let _createElement = React.createElement;

function createElement(type, config, ...children) {
  config = {...config, __originalConfig: config || {}};
  let element = _createElement(type, config, ...children);
  return element;
}

React.createElement = createElement;

