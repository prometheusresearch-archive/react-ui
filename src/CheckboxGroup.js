/**
 * @copyright 2016-present, Prometheus Research, LLC
 */

import invariant from 'invariant';
import React from 'react';

import Checkbox from './Checkbox';

export let primitiveValueStrategy = {

  findIndex(value, option) {
    if (!value) {
      return -1;
    }
    return value.indexOf(option.id);
  },

  optionToValue(option) {
    return option.id;
  },

  isChecked(value, option) {
    return this.findIndex(value, option) > -1;
  },

  update(value, option, checked) {
    value = value || [];
    value = value.slice(0);
    let idx = this.findIndex(value, option);
    if (checked) {
      invariant(
        idx === -1,
        'Duplicate id added'
      );
      value.push(this.optionToValue(option));
    } else {
      invariant(
        idx > -1,
        'Non-existent id unchecked'
      );
      value.splice(idx, 1);
    }
    return value;
  }

};

export default class CheckboxGroup extends React.Component {

  static propTypes = {
    valueStrategy: React.PropTypes.object
  };

  static defaultProps = {
    valueStrategy: primitiveValueStrategy
  };

  static stylesheet = {
    Root: 'div',
    CheckboxWrapper: 'div',
    Checkbox: Checkbox,
  };

  render() {
    let {options} = this.props;
    let {Root} = this.constructor.stylesheet;
    options = options.map(this.renderOption, this);
    return (
      <Root>
        {options}
      </Root>
    );
  }

  renderOption(option) {
    let {CheckboxWrapper, Checkbox} = this.constructor.stylesheet;
    let checked = this.props.valueStrategy.isChecked(this.props.value, option);
    return (
      <CheckboxWrapper key={option.id}>
        <Checkbox
          label={option.title}
          value={checked}
          onChange={this.onChange.bind(this, option)}
          />
      </CheckboxWrapper>
    );
  }

  onChange(option, checked) {
    let {value, valueStrategy} = this.props;
    value = valueStrategy.update(value, option, checked);
    this.props.onChange(value);
  }
}
