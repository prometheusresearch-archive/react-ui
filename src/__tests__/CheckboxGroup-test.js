/**
 * @copyright 2016, Prometheus Research, LLC
 */

import React from 'react';

import {createRenderer, spy, assert} from '../testutils';
import Checkbox from '../Checkbox';
import CheckboxGroup from '../CheckboxGroup';

describe('react-ui', function() {

  let options = [
    {id: 'a', title: 'A'},
    {id: 'b', title: 'B'}
  ];

  let renderer = null;

  beforeEach(function() {
    renderer = createRenderer();
  });

  describe('<CheckboxGroup />', function() {

    it('renders', function() {
      let onChange = spy();
      renderer.render(
        <CheckboxGroup
          value={['a']}
          onChange={onChange}
          options={options}
          />
      );
      let checkboxes = renderer.findAllWithType(Checkbox);

      assert(checkboxes.length === 2);
      assert(checkboxes[0].props.value);
      assert(!checkboxes[1].props.value);

      checkboxes[0].props.onChange(false);
      assert(onChange.callCount === 1);
      assert.deepEqual(onChange.lastCall.args[0], []);

      checkboxes[1].props.onChange(true);
      assert(onChange.callCount === 2);
      assert.deepEqual(onChange.lastCall.args[0], ['a', 'b']);
    });
  });

});
