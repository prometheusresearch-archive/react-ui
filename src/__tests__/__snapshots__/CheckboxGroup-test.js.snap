exports[`test renders 1`] = `
<div
  className="div--213234647"
  onBlur={[Function anonymous]}
  onFocus={[Function anonymous]}
  onKeyDown={[Function anonymous]}
  tabIndex={0}>
  <div
    className="div--177721320 div--vertical--177721320">
    <div
      title={undefined}
      variant={
        Object {
          "disabled": undefined,
          "ltr": true,
          "rtl": false
        }
      }>
      <input
        aria-labelledby="aria1"
        checked={true}
        className="input--125707658"
        disabled={undefined}
        inputRef={[Function anonymous]}
        onBlur={undefined}
        onChange={[Function anonymous]}
        onFocus={undefined}
        tabIndex={-1}
        type="checkbox" />
      <div
        className="Block--671161322 Block--ltr--671161322"
        onClick={[Function anonymous]}
        style={
          Object {
            "bottom": undefined,
            "display": undefined,
            "float": undefined,
            "height": undefined,
            "left": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "maxHeight": undefined,
            "maxWidth": undefined,
            "minHeight": undefined,
            "minWidth": undefined,
            "paddingBottom": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "position": "relative",
            "right": undefined,
            "textAlign": undefined,
            "top": undefined,
            "verticalAlign": undefined,
            "whiteSpace": undefined,
            "width": undefined
          }
        }>
        <div
          className="div--2240660900"
          id="aria1">
          A
        </div>
        <div
          className="div--3582280542" />
      </div>
    </div>
  </div>
  <div
    className="div--177721320 div--vertical--177721320">
    <div
      title={undefined}
      variant={
        Object {
          "disabled": undefined,
          "ltr": true,
          "rtl": false
        }
      }>
      <input
        aria-labelledby="aria2"
        checked={false}
        className="input--125707658"
        disabled={undefined}
        inputRef={[Function anonymous]}
        onBlur={undefined}
        onChange={[Function anonymous]}
        onFocus={undefined}
        tabIndex={-1}
        type="checkbox" />
      <div
        className="Block--671161322 Block--ltr--671161322"
        onClick={[Function anonymous]}
        style={
          Object {
            "bottom": undefined,
            "display": undefined,
            "float": undefined,
            "height": undefined,
            "left": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "maxHeight": undefined,
            "maxWidth": undefined,
            "minHeight": undefined,
            "minWidth": undefined,
            "paddingBottom": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "position": "relative",
            "right": undefined,
            "textAlign": undefined,
            "top": undefined,
            "verticalAlign": undefined,
            "whiteSpace": undefined,
            "width": undefined
          }
        }>
        <div
          className="div--2240660900"
          id="aria2">
          B
        </div>
        <div
          className="div--3582280542" />
      </div>
    </div>
  </div>
</div>
`;
